File: src/main/java/me/zeroeightsix/kami/module/modules/render/StorageESP.java
package me.zeroeightsix.kami.module.modules.render;

import me.zeroeightsix.kami.event.events.RenderEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import me.zeroeightsix.kami.util.ColourUtils;
import me.zeroeightsix.kami.util.GeometryMasks;
import me.zeroeightsix.kami.util.KamiTessellator;
import me.zeroeightsix.kami.util.Wrapper;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityItemFrame;
import net.minecraft.entity.item.EntityMinecartChest;
import net.minecraft.item.ItemShulkerBox;
import net.minecraft.tileentity.*;
import net.minecraft.util.math.BlockPos;
import org.lwjgl.opengl.GL11;

import java.util.ArrayList;

/**
 * Created by 086 on 10/12/2017.
 * Updated by S-B99 on 14/12/19
 */
@Module.Info(name = "StorageESP", description = "Draws nice little lines around storage items", category = Module.Category.RENDER)
public class StorageESP extends Module {

    private Setting<Boolean> chest = register(Settings.b("Chest", true));
    private Setting<Boolean> dispenser = register(Settings.b("Dispenser", true));
    private Setting<Boolean> shulker = register(Settings.b("Shulker", true));
    private Setting<Boolean> echest = register(Settings.b("Ender Chest", true));
    private Setting<Boolean> furnace = register(Settings.b("Furnace", true));
    private Setting<Boolean> hopper = register(Settings.b("Hopper", true));
    private Setting<Boolean> cart = register(Settings.b("Minecart", true));
    private Setting<Boolean> frame = register(Settings.b("Item Frame", true));


    private int getTileEntityColor(TileEntity tileEntity) {
        if (tileEntity instanceof TileEntityChest || tileEntity instanceof TileEntityDispenser)
            return ColourUtils.Colors.ORANGE;
        else if (tileEntity instanceof TileEntityShulkerBox)
            return ColourUtils.Colors.RED;
        else if (tileEntity instanceof TileEntityEnderChest)
            return ColourUtils.Colors.PURPLE;
        else if (tileEntity instanceof TileEntityFurnace)
            return ColourUtils.Colors.GRAY;
        else if (tileEntity instanceof TileEntityHopper)
            return ColourUtils.Colors.DARK_RED;
        else
            return -1;
    }

    private int getEntityColor(Entity entity) {
        if (entity instanceof EntityMinecartChest)
            return ColourUtils.Colors.ORANGE;
        else if (entity instanceof EntityItemFrame &&
                ((EntityItemFrame) entity).getDisplayedItem().getItem() instanceof ItemShulkerBox)
            return ColourUtils.Colors.YELLOW;
        else if (entity instanceof EntityItemFrame &&
                (!(((EntityItemFrame) entity).getDisplayedItem().getItem() instanceof ItemShulkerBox)))
            return ColourUtils.Colors.ORANGE;
        else
            return -1;
    }

    @Override
    public void onWorldRender(RenderEvent event) {
        ArrayList<Triplet<BlockPos, Integer, Integer>> a = new ArrayList<>();
        GlStateManager.pushMatrix();

        for (TileEntity tileEntity : Wrapper.getWorld().loadedTileEntityList) {
            BlockPos pos = tileEntity.getPos();
            int color = getTileEntityColor(tileEntity);
            int side = GeometryMasks.Quad.ALL;
            if (tileEntity instanceof TileEntityChest) {
                TileEntityChest chest = (TileEntityChest) tileEntity;
                // Leave only the colliding face and then flip the bits (~) to have ALL but that face
                if (chest.adjacentChestZNeg != null) side = ~(side & GeometryMasks.Quad.NORTH);
                if (chest.adjacentChestXPos != null) side = ~(side & GeometryMasks.Quad.EAST);
                if (chest.adjacentChestZPos != null) side = ~(side & GeometryMasks.Quad.SOUTH);
                if (chest.adjacentChestXNeg != null) side = ~(side & GeometryMasks.Quad.WEST);
            }
            if ((tileEntity instanceof TileEntityChest && chest.getValue()) || (tileEntity instanceof TileEntityDispenser && dispenser.getValue()) || (tileEntity instanceof TileEntityShulkerBox && shulker.getValue()) || (tileEntity instanceof TileEntityEnderChest && echest.getValue()) || (tileEntity instanceof TileEntityFurnace && furnace.getValue()) || (tileEntity instanceof TileEntityHopper && hopper.getValue()))
                if (color != -1)
                a.add(new Triplet<>(pos, color, side)); //GeometryTessellator.drawCuboid(event.getBuffer(), pos, GeometryMasks.Line.ALL, color);
        }

        for (Entity entity : Wrapper.getWorld().loadedEntityList) {
            BlockPos pos = entity.getPosition();
            int color = getEntityColor(entity);
            if ((entity instanceof EntityItemFrame && frame.getValue()) || (entity instanceof EntityMinecartChest && cart.getValue()))
                if (color != -1)
                    a.add(new Triplet<>(entity instanceof EntityItemFrame ? pos.add(0, -1, 0) : pos, color, GeometryMasks.Quad.ALL)); //GeometryTessellator.drawCuboid(event.getBuffer(), entity instanceof EntityItemFrame ? pos.add(0, -1, 0) : pos, GeometryMasks.Line.ALL, color);
        }

        KamiTessellator.prepare(GL11.GL_QUADS);
        for (Triplet<BlockPos, Integer, Integer> pair : a)
            KamiTessellator.drawBox(pair.getFirst(), changeAlpha(pair.getSecond(), 100), pair.getThird());
        KamiTessellator.release();

        GlStateManager.popMatrix();
        GlStateManager.enableTexture2D();
    }

    int changeAlpha(int origColor, int userInputedAlpha) {
        origColor = origColor & 0x00ffffff; //drop the previous alpha value
        return (userInputedAlpha << 24) | origColor; //add the one the user inputted
    }

    public class Triplet<T, U, V> {

        private final T first;
        private final U second;
        private final V third;

        public Triplet(T first, U second, V third) {
            this.first = first;
            this.second = second;
            this.third = third;
        }

        public T getFirst() {
            return first;
        }

        public U getSecond() {
            return second;
        }

        public V getThird() {
            return third;
        }
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/render/StorageESP.kt
package me.zeroeightsix.kami.module.modules.render

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import me.zeroeightsix.kami.event.SafeClientEvent
import me.zeroeightsix.kami.event.events.RenderWorldEvent
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.ModuleConfig.setting
import me.zeroeightsix.kami.util.color.ColorHolder
import me.zeroeightsix.kami.util.color.DyeColors
import me.zeroeightsix.kami.util.color.HueCycler
import me.zeroeightsix.kami.util.graphics.ESPRenderer
import me.zeroeightsix.kami.util.graphics.GeometryMasks
import me.zeroeightsix.kami.util.threads.safeAsyncListener
import net.minecraft.entity.Entity
import net.minecraft.entity.item.*
import net.minecraft.item.ItemShulkerBox
import net.minecraft.tileentity.*
import net.minecraft.util.math.AxisAlignedBB
import net.minecraftforge.fml.common.gameevent.TickEvent
import org.kamiblue.event.listener.listener

object StorageESP : Module(
    name = "StorageESP",
    description = "Draws an ESP on top of storage units",
    category = Category.RENDER
) {
    private val page by setting("Page", Page.TYPE)

    /* Type settings */
    private val chest by setting("Chest", true, { page == Page.TYPE })
    private val shulker by setting("Shulker", true, { page == Page.TYPE })
    private val enderChest by setting("EnderChest", true, { page == Page.TYPE })
    private val frame by setting("ItemFrame", true, { page == Page.TYPE })
    private val withShulkerOnly by setting("WithShulkerOnly", true, { page == Page.TYPE && frame })
    private val furnace by setting("Furnace", false, { page == Page.TYPE })
    private val dispenser by setting("Dispenser", false, { page == Page.TYPE })
    private val hopper by setting("Hopper", false, { page == Page.TYPE })
    private val cart by setting("Minecart", false, { page == Page.TYPE })

    /* Color settings */
    private val colorChest by setting("ChestColor", DyeColors.ORANGE, { page == Page.COLOR })
    private val colorDispenser by setting("DispenserColor", DyeColors.LIGHT_GRAY, { page == Page.COLOR })
    private val colorShulker by setting("ShulkerColor", DyeColors.MAGENTA, { page == Page.COLOR })
    private val colorEnderChest by setting("EnderChestColor", DyeColors.PURPLE, { page == Page.COLOR })
    private val colorFurnace by setting("FurnaceColor", DyeColors.LIGHT_GRAY, { page == Page.COLOR })
    private val colorHopper by setting("HopperColor", DyeColors.GRAY, { page == Page.COLOR })
    private val colorCart by setting("CartColor", DyeColors.GREEN, { page == Page.COLOR })
    private val colorFrame by setting("FrameColor", DyeColors.ORANGE, { page == Page.COLOR })

    /* Render settings */
    private val filled by setting("Filled", true, { page == Page.RENDER })
    private val outline by setting("Outline", true, { page == Page.RENDER })
    private val tracer by setting("Tracer", false, { page == Page.RENDER })
    private val aFilled by setting("FilledAlpha", 31, 0..255, 1, { page == Page.RENDER && filled })
    private val aOutline by setting("OutlineAlpha", 127, 0..255, 1, { page == Page.RENDER && outline })
    private val aTracer by setting("TracerAlpha", 200, 0..255, 1, { page == Page.RENDER && tracer })
    private val thickness by setting("LineThickness", 2.0f, 0.25f..5.0f, 0.25f, { page == Page.RENDER })

    private enum class Page {
        TYPE, COLOR, RENDER
    }

    override fun getHudInfo(): String {
        return renderer.size.toString()
    }

    private var cycler = HueCycler(600)
    private val renderer = ESPRenderer()

    init {
        listener<RenderWorldEvent> {
            renderer.render(false)
        }

        safeAsyncListener<TickEvent.ClientTickEvent> {
            if (it.phase != TickEvent.Phase.START) return@safeAsyncListener

            cycler++
            renderer.clear()
            val cached = ArrayList<Triple<AxisAlignedBB, ColorHolder, Int>>()

            coroutineScope {
                launch(Dispatchers.Default) {
                    updateRenderer()
                }
                launch(Dispatchers.Default) {
                    updateTileEntities(cached)
                }
                launch(Dispatchers.Default) {
                    updateEntities(cached)
                }
            }

            renderer.replaceAll(cached)
        }
    }

    private fun updateRenderer() {
        renderer.aFilled = if (filled) aFilled else 0
        renderer.aOutline = if (outline) aOutline else 0
        renderer.aTracer = if (tracer) aTracer else 0
        renderer.thickness = thickness
    }

    private fun SafeClientEvent.updateTileEntities(list: MutableList<Triple<AxisAlignedBB, ColorHolder, Int>>) {
        for (tileEntity in world.loadedTileEntityList.toList()) {
            if (!checkTileEntityType(tileEntity)) continue

            val box = world.getBlockState(tileEntity.pos).getSelectedBoundingBox(world, tileEntity.pos) ?: continue
            val color = getTileEntityColor(tileEntity) ?: continue
            var side = GeometryMasks.Quad.ALL

            if (tileEntity is TileEntityChest) {
                // Leave only the colliding face and then flip the bits (~) to have ALL but that face
                if (tileEntity.adjacentChestZNeg != null) side = (side and GeometryMasks.Quad.NORTH).inv()
                if (tileEntity.adjacentChestXPos != null) side = (side and GeometryMasks.Quad.EAST).inv()
                if (tileEntity.adjacentChestZPos != null) side = (side and GeometryMasks.Quad.SOUTH).inv()
                if (tileEntity.adjacentChestXNeg != null) side = (side and GeometryMasks.Quad.WEST).inv()
            }

            list.add(Triple(box, color, side))
        }
    }

    private fun checkTileEntityType(tileEntity: TileEntity) =
        chest && tileEntity is TileEntityChest
            || dispenser && tileEntity is TileEntityDispenser
            || shulker && tileEntity is TileEntityShulkerBox
            || enderChest && tileEntity is TileEntityEnderChest
            || furnace && tileEntity is TileEntityFurnace
            || hopper && tileEntity is TileEntityHopper

    private fun getTileEntityColor(tileEntity: TileEntity): ColorHolder? {
        val color = when (tileEntity) {
            is TileEntityChest -> colorChest
            is TileEntityDispenser -> colorDispenser
            is TileEntityShulkerBox -> colorShulker
            is TileEntityEnderChest -> colorEnderChest
            is TileEntityFurnace -> colorFurnace
            is TileEntityHopper -> colorHopper
            else -> return null
        }.color
        return if (color == DyeColors.RAINBOW.color) {
            cycler.currentRgb()
        } else color
    }

    private fun SafeClientEvent.updateEntities(list: MutableList<Triple<AxisAlignedBB, ColorHolder, Int>>) {
        for (entity in world.loadedEntityList.toList()) {
            if (!checkEntityType(entity)) continue

            val box = entity.renderBoundingBox ?: continue
            val color = getEntityColor(entity) ?: continue

            list.add(Triple(box, color, GeometryMasks.Quad.ALL))
        }
    }

    private fun checkEntityType(entity: Entity) =
        entity is EntityItemFrame && frameShulkerOrAny(entity)
            || (entity is EntityMinecartChest || entity is EntityMinecartHopper || entity is EntityMinecartFurnace) && cart

    private fun getEntityColor(entity: Entity): ColorHolder? {
        val color = when (entity) {
            is EntityMinecartContainer -> colorCart
            is EntityItemFrame -> colorFrame
            else -> return null
        }.color
        return if (color == DyeColors.RAINBOW.color) {
            cycler.currentRgb()
        } else color
    }

    private fun frameShulkerOrAny(entity: EntityItemFrame) =
        frame && (!withShulkerOnly || entity.displayedItem.item is ItemShulkerBox)
}
